### ISSUE TITLE
Hard lock/crash when searching for '$' character in article

### ISSUE THREAD

**Issue by apetresc** [newsboat/newsboat#313]
**Newsboat version (copy from `newsboat -v`)**:
```
newsboat 2.13.0 - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2018 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2018 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `newsboat -vv' to see the full text.)
It bundles JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json
It bundles an alphanum algorithm implementation licensed under the MIT license: http://www.davekoelle.com/alphanum.html

newsboat 2.13.0
System: Linux 4.18.11-1-ck-haswell (x86_64)
Compiler: g++ 8.2.1 20180831
ncurses: ncurses 6.1.20180127 (compiled with 6.1)
libcurl: libcurl/7.61.1 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.4) libssh2/1.8.0 nghttp2/1.33.0 (compiled with 7.61.1)
SQLite: 3.25.2 (compiled with 3.25.2)
libxml2: compiled with 2.9.8
```

Steps to reproduce the issue:

1. Open up an article (any article seems to work)
2. Do a text search for the `$` character
3. newsboat will immediately shoot up to 100% CPU usage and hang forever. It does not respond to any other keys (like to exit the article view), nor does Ctrl-C do anything. I have to `kill` it from another shell.

This seems to be 100% reproducible on all my machines, so it's not some tricky race condition or anything.

**Comment by kmws** [comment]
Reproduced with [`acc367`](https://github.com/newsboat/newsboat/commit/acc367182ff2818fcfbf1a377ccbbaa1c86052b5).

**Comment by zaowen** [comment]
I've figured out why this is happening but need some advice on how to proceed.

The problem is `regexec()` treats `$` as a zero length string just before the newline. Since we just keep searching until `regexec()` returns an error `quote_and_highlight()` spins adding `<0></>` over and over again. This is an easy fix but how depends on desired functionality.
1. Just highlight nothing
2. Vim will highlight an inserted space character
3. Some other third option
I like vim so I like 2, thoughts?

The problem is this bug also occurs on searches for `^` which matches a zero length string at the begining of the line. This is a less easy fix but how depends on desired functionality.
1. Just highlight nothing
2. Vim will highlight the first character of the line
3. Some other third option
I dont like any of these options.

**Comment by Minoru** [comment]
@apetresc, I forgot to ask—*why* did you search for line endings? If there was an intent behind it, I'd like to know :)

**Comment by Minoru** [comment]
@zaowen, I think it's better to follow Mutt here. It implements first option, and doesn't highlight anything.

I don't see a reason why one would search just for line beginning or ending, so I can't tell how useful highlighting will be for such user. Let's wait for a reply from Adrian.

**Comment by zaowen** [comment]
@Minoru I just found that it also exhibits the same behavior with `^$`. I'm going to do a bit looking around for other emtpy string valid regexes so the PR can wipe out all variations on this bug.

**Comment by Minoru** [comment]
> I'm going to do a bit looking around for other emtpy string valid regexes

That might be useful for tests or just for curiosity's sake, but otherwise I don't think you need to do that. I think this issue can be fixed once and for all by exiting the loop if regex matched an empty string. Do you see any flaws in this solution?

**Comment by zaowen** [comment]
That solution would return no matches on something like `q*` Im not sure if we care though.

I think a better solution would be to step over emtpy matched searches

I'm looking into how mutt handles searching to see if that logic can be copied wholesale.

**Comment by apetresc** [comment]
(In terms of the big picture, it might be worth it to have separate plain search and regex search functions, and keep the default article search bound to the plain one. I'm sure someone somewhere wants to regex search their articles but I think it's much more common that someone tries to do a plaintext search involving a regex metachar like `$`, `^`, or `[` and ends up with confusing incorrect results.)

**Comment by Minoru** [comment]
@apetresc It is definitely worthy of reporing, no argument there! I was just curious if you had some specific goal in mind. Thanks :)

Two search modes will be harder to discover and will take up more keys to bind. Also there were no user requests to that effect yet, so it seems like no-one is inconvenienced by regexes. Maybe we could just change "Search for:" to "Search for regex:" to prevent the confusion?

@zaowen Well, if the match is empty, there is nothing to highlight. We'll still highlight "q", "qq" and so forth, because they are non-empty. Seems okay to me.

You're absolutely right about step-over; I didn't think of that. An empty match in the middle of the string shouldn't stop us from processing the rest of it.

Just in case, note that we can't copy *code* from Mutt; it's under GPLv2 and we're under MIT license.

**Comment by zaowen** [comment]
> Just in case, note that we can't copy code from Mutt; it's under GPLv2 and we're under MIT license.

Just to be clear I do know this.

**Comment by zaowen** [comment]
Should this be closed?

**Comment by Minoru** [comment]
Oh, right, it should be! GitHub didn't auto-close this because "Fixes #313" was only present in the PR, not the commit message, and I merged the commits manually (not via GitHub). My bad, closing now.
